File: backend/app/services/hrm_service.py
"""
Comprehensive HRM Service
"""
from datetime import date, datetime, timedelta
from typing import List, Optional, Dict, Any

from sqlalchemy import select, func, and_, or_, desc
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload
from uuid import UUID

from app.models.hrm_models import (
from app.schemas.hrm.hrm_schemas import (


    Employee, Department, LeaveRequest, AttendanceRecord, 
    PerformanceReview, TrainingRecord, Policy, JobOpening, 
    Candidate, Interview
)
    EmployeeCreate, EmployeeUpdate, DepartmentCreate,
    LeaveRequestCreate, AttendanceRecordCreate, 
    PerformanceReviewCreate, PolicyCreate
)

class HRMService:
    """Comprehensive HRM Service with real-time data integration"""
    
    def __init__(self):
        """  Init  ."""
        self.mock_tenant_id = UUID("12345678-1234-5678-9012-123456789012")
    
    # Employee Management
    async def create_employee(
        self, 
        db: AsyncSession, 
        employee_data: EmployeeCreate,
        tenant_id: UUID = None
    ) -> Employee:
        """Create Employee."""
        """Create a new employee"""
        tenant_id = tenant_id or self.mock_tenant_id
        
        employee = Employee(
            tenant_id=tenant_id,
            employee_id=employee_data.employee_id,
            first_name=employee_data.first_name,
            middle_name=employee_data.middle_name,
            last_name=employee_data.last_name,
            email=employee_data.email,
            phone_number=employee_data.phone_number,
            date_of_birth=employee_data.date_of_birth,
            gender=employee_data.gender,
            job_title=employee_data.job_title,
            department_id=employee_data.department_id,
            hire_date=employee_data.hire_date,
            employment_type=employee_data.employment_type,
            base_salary=employee_data.base_salary,
            manager_id=employee_data.manager_id
        )
        
        db.add(employee)
        await db.commit()
        await db.refresh(employee)
        return employee
    
    async def get_employees(
        self,
        db: AsyncSession,
        tenant_id: UUID = None,
        skip: int = 0,
        limit: int = 100,
        department_id: UUID = None,
        is_active: bool = True,
        search: str = None
    ) -> List[Employee]:
        """Get Employees."""
        """Get employees with filters"""
        tenant_id = tenant_id or self.mock_tenant_id
        
        query = select(Employee).options(
            selectinload(Employee.department),
            selectinload(Employee.manager)
        ).where(Employee.tenant_id == tenant_id)
        
        if is_active is not None:
            query = query.where(Employee.is_active == is_active)
        
        if department_id:
            query = query.where(Employee.department_id == department_id)
        
        if search:
            search_filter = or_(
                Employee.first_name.ilike(f"%{search}%"),
                Employee.last_name.ilike(f"%{search}%"),
                Employee.email.ilike(f"%{search}%"),
                Employee.employee_id.ilike(f"%{search}%")
            )
            query = query.where(search_filter)
        
        query = query.offset(skip).limit(limit)
        result = await db.execute(query)
        return result.scalars().all()
    
    async def get_employee_by_id(
        self, 
        db: AsyncSession, 
        employee_id: UUID,
        tenant_id: UUID = None
    ) -> Optional[Employee]:
        """Get Employee By Id."""
        """Get employee by ID"""
        tenant_id = tenant_id or self.mock_tenant_id
        
        query = select(Employee).options(
            selectinload(Employee.department),
            selectinload(Employee.manager),
            selectinload(Employee.direct_reports)
        ).where(
            and_(Employee.id == employee_id, Employee.tenant_id == tenant_id)
        )
        
        result = await db.execute(query)
        return result.scalar_one_or_none()
    
    async def update_employee(
        self,
        db: AsyncSession,
        employee_id: UUID,
        employee_data: EmployeeUpdate,
        tenant_id: UUID = None
    ) -> Optional[Employee]:
        """Update Employee."""
        """Update employee"""
        employee = await self.get_employee_by_id(db, employee_id, tenant_id)
        if not employee:
            return None
        
        for field, value in employee_data.dict(exclude_unset=True).items():
            setattr(employee, field, value)
        
        employee.updated_at = datetime.utcnow()
        await db.commit()
        await db.refresh(employee)
        return employee
    
    # Department Management
    async def create_department(
        self,
        db: AsyncSession,
        department_data: DepartmentCreate,
        tenant_id: UUID = None
    ) -> Department:
        """Create Department."""
        """Create a new department"""
        tenant_id = tenant_id or self.mock_tenant_id
        
        department = Department(
            tenant_id=tenant_id,
            name=department_data.name,
            description=department_data.description,
            manager_id=department_data.manager_id,
            parent_department_id=department_data.parent_department_id,
            budget=department_data.budget
        )
        
        db.add(department)
        await db.commit()
        await db.refresh(department)
        return department
    
    async def get_departments(
        self,
        db: AsyncSession,
        tenant_id: UUID = None,
        include_inactive: bool = False
    ) -> List[Department]:
        """Get Departments."""
        """Get all departments"""
        tenant_id = tenant_id or self.mock_tenant_id
        
        query = select(Department).options(
            selectinload(Department.manager),
            selectinload(Department.employees)
        ).where(Department.tenant_id == tenant_id)
        
        if not include_inactive:
            query = query.where(Department.is_active == True)
        
        result = await db.execute(query)
        return result.scalars().all()
    
    # Leave Management
    async def create_leave_request(
        self,
        db: AsyncSession,
        leave_data: LeaveRequestCreate,
        employee_id: UUID,
        tenant_id: UUID = None
    ) -> LeaveRequest:
        """Create Leave Request."""
        """Create a leave request"""
        tenant_id = tenant_id or self.mock_tenant_id
        
        # Calculate days
        days_requested = (leave_data.end_date - leave_data.start_date).days + 1
        
        leave_request = LeaveRequest(
            tenant_id=tenant_id,
            employee_id=employee_id,
            leave_type=leave_data.leave_type,
            start_date=leave_data.start_date,
            end_date=leave_data.end_date,
            days_requested=days_requested,
            reason=leave_data.reason
        )
        
        db.add(leave_request)
        await db.commit()
        await db.refresh(leave_request)
        return leave_request
    
    async def get_leave_requests(
        self,
        db: AsyncSession,
        tenant_id: UUID = None,
        employee_id: UUID = None,
        status: str = None,
        start_date: date = None,
        end_date: date = None
    ) -> List[LeaveRequest]:
        """Get Leave Requests."""
        """Get leave requests with filters"""
        tenant_id = tenant_id or self.mock_tenant_id
        
        query = select(LeaveRequest).options(
            selectinload(LeaveRequest.employee),
            selectinload(LeaveRequest.approver)
        ).where(LeaveRequest.tenant_id == tenant_id)
        
        if employee_id:
            query = query.where(LeaveRequest.employee_id == employee_id)
        
        if status:
            query = query.where(LeaveRequest.status == status)
        
        if start_date:
            query = query.where(LeaveRequest.start_date >= start_date)
        
        if end_date:
            query = query.where(LeaveRequest.end_date <= end_date)
        
        query = query.order_by(desc(LeaveRequest.created_at))
        result = await db.execute(query)
        return result.scalars().all()
    
    async def approve_leave_request(
        self,
        db: AsyncSession,
        request_id: UUID,
        approved_by: UUID,
        tenant_id: UUID = None
    ) -> Optional[LeaveRequest]:
        """Approve Leave Request."""
        """Approve a leave request"""
        tenant_id = tenant_id or self.mock_tenant_id
        
        query = select(LeaveRequest).where(
            and_(
                LeaveRequest.id == request_id,
                LeaveRequest.tenant_id == tenant_id
            )
        )
        result = await db.execute(query)
        leave_request = result.scalar_one_or_none()
        
        if leave_request:
            leave_request.status = "APPROVED"
            leave_request.approved_by = approved_by
            leave_request.approved_at = datetime.utcnow()
            await db.commit()
            await db.refresh(leave_request)
        
        return leave_request
    
    # Attendance Management
    async def record_attendance(
        self,
        db: AsyncSession,
        attendance_data: AttendanceRecordCreate,
        employee_id: UUID,
        tenant_id: UUID = None
    ) -> AttendanceRecord:
        """Record Attendance."""
        """Record attendance"""
        tenant_id = tenant_id or self.mock_tenant_id
        
        # Check if record already exists for this date
        existing_query = select(AttendanceRecord).where(
            and_(
                AttendanceRecord.employee_id == employee_id,
                AttendanceRecord.date == attendance_data.date,
                AttendanceRecord.tenant_id == tenant_id
            )
        )
        result = await db.execute(existing_query)
        existing_record = result.scalar_one_or_none()
        
        if existing_record:
            # Update existing record
            for field, value in attendance_data.dict(exclude_unset=True).items():
                setattr(existing_record, field, value)
            
            # Calculate total hours if both check-in and check-out are present
            if existing_record.check_in_time and existing_record.check_out_time:
                duration = existing_record.check_out_time - existing_record.check_in_time
                total_hours = duration.total_seconds() / 3600
                if existing_record.break_duration:
                    total_hours -= existing_record.break_duration / 60
                existing_record.total_hours = round(total_hours, 2)
            
            await db.commit()
            await db.refresh(existing_record)
            return existing_record
        else:
            # Create new record
            attendance = AttendanceRecord(
                tenant_id=tenant_id,
                employee_id=employee_id,
                **attendance_data.dict()
            )
            
            db.add(attendance)
            await db.commit()
            await db.refresh(attendance)
            return attendance
    
    async def get_attendance_records(
        self,
        db: AsyncSession,
        tenant_id: UUID = None,
        employee_id: UUID = None,
        start_date: date = None,
        end_date: date = None
    ) -> List[AttendanceRecord]:
        """Get Attendance Records."""
        """Get attendance records"""
        tenant_id = tenant_id or self.mock_tenant_id
        
        query = select(AttendanceRecord).options(
            selectinload(AttendanceRecord.employee)
        ).where(AttendanceRecord.tenant_id == tenant_id)
        
        if employee_id:
            query = query.where(AttendanceRecord.employee_id == employee_id)
        
        if start_date:
            query = query.where(AttendanceRecord.date >= start_date)
        
        if end_date:
            query = query.where(AttendanceRecord.date <= end_date)
        
        query = query.order_by(desc(AttendanceRecord.date))
        result = await db.execute(query)
        return result.scalars().all()
    
    # Performance Management
    async def create_performance_review(
        self,
        db: AsyncSession,
        review_data: PerformanceReviewCreate,
        employee_id: UUID,
        reviewer_id: UUID,
        tenant_id: UUID = None
    ) -> PerformanceReview:
        """Create Performance Review."""
        """Create performance review"""
        tenant_id = tenant_id or self.mock_tenant_id
        
        review = PerformanceReview(
            tenant_id=tenant_id,
            employee_id=employee_id,
            reviewer_id=reviewer_id,
            **review_data.dict()
        )
        
        db.add(review)
        await db.commit()
        await db.refresh(review)
        return review
    
    async def get_performance_reviews(
        self,
        db: AsyncSession,
        tenant_id: UUID = None,
        employee_id: UUID = None,
        year: int = None
    ) -> List[PerformanceReview]:
        """Get Performance Reviews."""
        """Get performance reviews"""
        tenant_id = tenant_id or self.mock_tenant_id
        
        query = select(PerformanceReview).options(
            selectinload(PerformanceReview.employee),
            selectinload(PerformanceReview.reviewer)
        ).where(PerformanceReview.tenant_id == tenant_id)
        
        if employee_id:
            query = query.where(PerformanceReview.employee_id == employee_id)
        
        if year:
            start_date = date(year, 1, 1)
            end_date = date(year, 12, 31)
            query = query.where(
                and_(
                    PerformanceReview.review_date >= start_date,
                    PerformanceReview.review_date <= end_date
                )
            )
        
        query = query.order_by(desc(PerformanceReview.review_date))
        result = await db.execute(query)
        return result.scalars().all()
    
    # Analytics and Reporting
    async def get_hr_analytics(
        self,
        db: AsyncSession,
        tenant_id: UUID = None
    ) -> Dict[str, Any]:
        """Get Hr Analytics."""
        """Get HR analytics dashboard data"""
        tenant_id = tenant_id or self.mock_tenant_id
        
        # Employee counts
        total_employees_query = select(func.count(Employee.id)).where(
            Employee.tenant_id == tenant_id
        )
        active_employees_query = select(func.count(Employee.id)).where(
            and_(Employee.tenant_id == tenant_id, Employee.is_active == True)
        )
        
        # Leave requests
        pending_leave_query = select(func.count(LeaveRequest.id)).where(
            and_(
                LeaveRequest.tenant_id == tenant_id,
                LeaveRequest.status == "PENDING"
            )
        )
        
        # Department breakdown
        dept_breakdown_query = select(
            Department.name,
            func.count(Employee.id).label('employee_count')
        ).select_from(
            Department
        ).outerjoin(Employee).where(
            and_(
                Department.tenant_id == tenant_id,
                Department.is_active == True
            )
        ).group_by(Department.name)
        
        # Recent hires (last 30 days)
        thirty_days_ago = date.today() - timedelta(days=30)
        recent_hires_query = select(Employee).where(
            and_(
                Employee.tenant_id == tenant_id,
                Employee.hire_date >= thirty_days_ago,
                Employee.is_active == True
            )
        ).order_by(desc(Employee.hire_date)).limit(10)
        
        # Execute queries
        total_employees = (await db.execute(total_employees_query)).scalar()
        active_employees = (await db.execute(active_employees_query)).scalar()
        pending_leave = (await db.execute(pending_leave_query)).scalar()
        dept_breakdown = (await db.execute(dept_breakdown_query)).all()
        recent_hires = (await db.execute(recent_hires_query)).scalars().all()
        
        return {
            "total_employees": total_employees or 0,
            "active_employees": active_employees or 0,
            "inactive_employees": (total_employees or 0) - (active_employees or 0),
            "pending_leave_requests": pending_leave or 0,
            "department_breakdown": [
                {"department": row.name, "count": row.employee_count}
                for row in dept_breakdown
            ],
            "recent_hires": [
                {
                    "name": emp.full_name,
                    "position": emp.job_title,
                    "hire_date": emp.hire_date.isoformat(),
                    "department": emp.department.name if emp.department else None
                }
                for emp in recent_hires
            ],
            "average_tenure_months": await self._calculate_average_tenure(db, tenant_id)
        }
    
    async def _calculate_average_tenure(
        self, 
        db: AsyncSession, 
        tenant_id: UUID
    ) -> float:
        """Calculate Average Tenure."""
        """Calculate average employee tenure in months"""
        query = select(Employee.hire_date).where(
            and_(Employee.tenant_id == tenant_id, Employee.is_active == True)
        )
        result = await db.execute(query)
        hire_dates = result.scalars().all()
        
        if not hire_dates:
            return 0.0
        
        today = date.today()
        total_months = sum(
            (today.year - hire_date.year) * 12 + (today.month - hire_date.month)
            for hire_date in hire_dates
        )
        
        return round(total_months / len(hire_dates), 1)
    
    # Policy Management
    async def create_policy(
        self,
        db: AsyncSession,
        policy_data: PolicyCreate,
        tenant_id: UUID = None
    ) -> Policy:
        """Create Policy."""
        """Create a new policy"""
        tenant_id = tenant_id or self.mock_tenant_id
        
        policy = Policy(
            tenant_id=tenant_id,
            **policy_data.dict()
        )
        
        db.add(policy)
        await db.commit()
        await db.refresh(policy)
        return policy
    
    async def get_policies(
        self,
        db: AsyncSession,
        tenant_id: UUID = None,
        category: str = None,
        status: str = None
    ) -> List[Policy]:
        """Get Policies."""
        """Get policies with filters"""
        tenant_id = tenant_id or self.mock_tenant_id
        
        query = select(Policy).where(Policy.tenant_id == tenant_id)
        
        if category:
            query = query.where(Policy.category == category)
        
        if status:
            query = query.where(Policy.status == status)
        
        query = query.order_by(desc(Policy.created_at))
        result = await db.execute(query)
        return result.scalars().all()

# Create service instance
hrm_service = HRMService()

File: frontend/src/services/hrmService.ts
import { api } from '@/utils/api'

// Comprehensive HRM Interfaces
export interface Employee {
  id?: string
  tenant_id?: string
  employee_id: string
  first_name: string
  middle_name?: string
  last_name: string
  full_name?: string
  email: string
  phone_number: string
  date_of_birth?: string
  gender?: string
  marital_status?: string
  national_id?: string
  job_title: string
  department_id?: string
  department?: Department
  manager_id?: string
  manager?: Employee
  hire_date: string
  termination_date?: string
  employment_type: string
  base_salary: number
  currency?: string
  is_active: boolean
  created_at?: string
  updated_at?: string
}

export interface Department {
  id: string
  tenant_id?: string
  name: string
  description?: string
  manager_id?: string
  manager?: Employee
  parent_department_id?: string
  parent_department?: Department
  budget?: number
  employee_count?: number
  is_active: boolean
  created_at?: string
  updated_at?: string
}

export interface LeaveRequest {
  id?: string
  tenant_id?: string
  employee_id: string
  employee?: Employee
  leave_type: string
  start_date: string
  end_date: string
  days_requested: number
  reason?: string
  status: string
  approved_by?: string
  approver?: Employee
  approved_at?: string
  rejection_reason?: string
  created_at?: string
  updated_at?: string
}

export interface AttendanceRecord {
  id?: string
  tenant_id?: string
  employee_id: string
  employee?: Employee
  date: string
  check_in_time?: string
  check_out_time?: string
  break_duration?: number
  total_hours?: number
  overtime_hours?: number
  status: string
  notes?: string
  created_at?: string
  updated_at?: string
}

export interface PerformanceReview {
  id?: string
  tenant_id?: string
  employee_id: string
  employee?: Employee
  reviewer_id: string
  reviewer?: Employee
  review_period_start: string
  review_period_end: string
  review_date: string
  overall_rating: string
  goals_achievement?: number
  strengths?: string
  areas_for_improvement?: string
  development_goals?: string
  manager_comments?: string
  employee_comments?: string
  status: string
  created_at?: string
  updated_at?: string
}

export interface TrainingRecord {
  id?: string
  tenant_id?: string
  employee_id: string
  employee?: Employee
  training_name: string
  training_type: string
  provider?: string
  start_date: string
  end_date?: string
  duration_hours?: number
  cost?: number
  status: string
  completion_date?: string
  certificate_url?: string
  notes?: string
  created_at?: string
  updated_at?: string
}

export interface Policy {
  id?: string
  tenant_id?: string
  title: string
  category: string
  description?: string
  content: string
  version: string
  effective_date: string
  expiry_date?: string
  status: string
  approval_required: boolean
  approved_by?: string
  approver?: Employee
  approved_at?: string
  created_at?: string
  updated_at?: string
  created_by?: string
  updated_by?: string
}

export interface JobOpening {
  id?: string
  tenant_id?: string
  title: string
  department_id?: string
  department?: Department
  description: string
  requirements?: string
  employment_type: string
  salary_min?: number
  salary_max?: number
  location?: string
  remote_allowed: boolean
  status: string
  posted_date: string
  closing_date?: string
  hiring_manager_id?: string
  hiring_manager?: Employee
  created_at?: string
  updated_at?: string
}

export interface Candidate {
  id?: string
  tenant_id?: string
  job_opening_id: string
  job_opening?: JobOpening
  first_name: string
  last_name: string
  email: string
  phone_number?: string
  resume_url?: string
  cover_letter?: string
  status: string
  source?: string
  applied_date: string
  created_at?: string
  updated_at?: string
}

export interface Interview {
  id?: string
  tenant_id?: string
  candidate_id: string
  candidate?: Candidate
  interviewer_id: string
  interviewer?: Employee
  interview_type: string
  scheduled_date: string
  duration_minutes?: number
  location?: string
  meeting_link?: string
  status: string
  feedback?: string
  rating?: number
  recommendation?: string
  created_at?: string
  updated_at?: string
}

export interface HRAnalytics {
  total_employees: number
  active_employees: number
  inactive_employees: number
  pending_leave_requests: number
  department_breakdown: Array<{
    department: string
    count: number
  }>
  recent_hires: Array<{
    name: string
    position: string
    hire_date: string
    department?: string
  }>
  average_tenure_months: number
}

export interface EmployeeDashboard {
  employee: Employee
  pending_leave_requests: LeaveRequest[]
  recent_attendance: AttendanceRecord[]
  upcoming_reviews: PerformanceReview[]
  training_progress: TrainingRecord[]
  team_members: Employee[]
}

class HRMService {
  private baseUrl = '/api/v1/hrm'

  // Employee Management
  async getEmployees(params?: {
    skip?: number
    limit?: number
    department_id?: string
    is_active?: boolean
    search?: string
  }): Promise<{ data: Employee[] }> {
    const response = await api.get('/api/v1/hrm/employees')
    return { data: response.data || [] }
  }

  async createEmployee(employee: any): Promise<Employee> {
    const response = await fetch('/api/v1/hrm/employees', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(employee)
    })
    return response.json()
  }

  async updateEmployee(id: string, employee: Partial<Employee>): Promise<Employee> {
    const response = await api.put(`${this.baseUrl}/employees/${id}`, employee)
    return response.data
  }

  async deleteEmployee(id: string): Promise<void> {
    await api.delete(`${this.baseUrl}/employees/${id}`)
  }

  async getEmployee(id: string): Promise<Employee> {
    const response = await api.get(`${this.baseUrl}/employees/${id}`)
    return response.data
  }

  // Department Management
  async getDepartments(params?: {
    include_inactive?: boolean
  }): Promise<{ data: Department[] }> {
    const response = await api.get('/api/v1/hrm/departments')
    return { data: response.data || [] }
  }

  async createDepartment(department: Omit<Department, 'id' | 'tenant_id' | 'created_at' | 'updated_at'>): Promise<Department> {
    const response = await api.post(`${this.baseUrl}/departments`, department)
    return response.data
  }

  async updateDepartment(id: string, department: Partial<Department>): Promise<Department> {
    const response = await api.put(`${this.baseUrl}/departments/${id}`, department)
    return response.data
  }

  async deleteDepartment(id: string): Promise<void> {
    await api.delete(`${this.baseUrl}/departments/${id}`)
  }

  // Leave Management
  async getLeaveRequests(params?: {
    employee_id?: string
    status?: string
    start_date?: string
    end_date?: string
  }): Promise<{ data: LeaveRequest[] }> {
    const response = await api.get(`${this.baseUrl}/leave-requests`, { params })
    return response
  }

  async createLeaveRequest(employeeId: string, request: Omit<LeaveRequest, 'id' | 'tenant_id' | 'employee_id' | 'employee' | 'days_requested' | 'status' | 'created_at' | 'updated_at'>): Promise<LeaveRequest> {
    const response = await api.post(`${this.baseUrl}/leave-requests?employee_id=${employeeId}`, request)
    return response.data
  }

  async approveLeaveRequest(id: string): Promise<LeaveRequest> {
    const response = await api.post(`${this.baseUrl}/leave-requests/${id}/approve`)
    return response.data
  }

  async rejectLeaveRequest(id: string, reason: string): Promise<LeaveRequest> {
    const response = await api.post(`${this.baseUrl}/leave-requests/${id}/reject`, { reason })
    return response.data
  }

  // Attendance Management
  async getAttendanceRecords(params?: {
    employee_id?: string
    start_date?: string
    end_date?: string
  }): Promise<{ data: AttendanceRecord[] }> {
    const response = await api.get(`${this.baseUrl}/attendance`, { params })
    return response
  }

  async recordAttendance(employeeId: string, record: Omit<AttendanceRecord, 'id' | 'tenant_id' | 'employee_id' | 'employee' | 'total_hours' | 'created_at' | 'updated_at'>): Promise<AttendanceRecord> {
    const response = await api.post(`${this.baseUrl}/attendance?employee_id=${employeeId}`, record)
    return response.data
  }

  // Performance Management
  async getPerformanceReviews(params?: {
    employee_id?: string
    year?: number
  }): Promise<{ data: PerformanceReview[] }> {
    const response = await api.get(`${this.baseUrl}/performance-reviews`, { params })
    return response
  }

  async createPerformanceReview(employeeId: string, reviewerId: string, review: Omit<PerformanceReview, 'id' | 'tenant_id' | 'employee_id' | 'reviewer_id' | 'employee' | 'reviewer' | 'status' | 'created_at' | 'updated_at'>): Promise<PerformanceReview> {
    const response = await api.post(`${this.baseUrl}/performance-reviews?employee_id=${employeeId}&reviewer_id=${reviewerId}`, review)
    return response.data
  }

  async updatePerformanceReview(id: string, review: Partial<PerformanceReview>): Promise<PerformanceReview> {
    const response = await api.put(`${this.baseUrl}/performance-reviews/${id}`, review)
    return response.data
  }

  async deletePerformanceReview(id: string): Promise<void> {
    await api.delete(`${this.baseUrl}/performance-reviews/${id}`)
  }

  // Training Management
  async getTrainingRecords(params?: {
    employee_id?: string
    status?: string
  }): Promise<{ data: TrainingRecord[] }> {
    const response = await api.get(`${this.baseUrl}/training-records`, { params })
    return response
  }

  async createTrainingRecord(employeeId: string, training: Omit<TrainingRecord, 'id' | 'tenant_id' | 'employee_id' | 'employee' | 'created_at' | 'updated_at'>): Promise<TrainingRecord> {
    const response = await api.post(`${this.baseUrl}/training-records?employee_id=${employeeId}`, training)
    return response.data
  }

  async updateTrainingRecord(id: string, training: Partial<TrainingRecord>): Promise<TrainingRecord> {
    const response = await api.put(`${this.baseUrl}/training-records/${id}`, training)
    return response.data
  }

  async deleteTrainingRecord(id: string): Promise<void> {
    await api.delete(`${this.baseUrl}/training-records/${id}`)
  }

  // Policy Management
  async getPolicies(params?: {
    category?: string
    status?: string
  }): Promise<{ data: Policy[] }> {
    const response = await api.get(`${this.baseUrl}/policies`, { params })
    return response
  }

  async createPolicy(policy: Omit<Policy, 'id' | 'tenant_id' | 'version' | 'approved_by' | 'approved_at' | 'created_at' | 'updated_at' | 'created_by' | 'updated_by'>): Promise<Policy> {
    const response = await api.post(`${this.baseUrl}/policies`, policy)
    return response.data
  }

  async updatePolicy(id: string, policy: Partial<Policy>): Promise<Policy> {
    const response = await api.put(`${this.baseUrl}/policies/${id}`, policy)
    return response.data
  }

  async deletePolicy(id: string): Promise<void> {
    await api.delete(`${this.baseUrl}/policies/${id}`)
  }

  // Recruitment Management
  async getJobOpenings(params?: {
    department_id?: string
    status?: string
  }): Promise<{ data: JobOpening[] }> {
    const response = await api.get(`${this.baseUrl}/job-openings`, { params })
    return response
  }

  async createJobOpening(jobOpening: Omit<JobOpening, 'id' | 'tenant_id' | 'created_at' | 'updated_at'>): Promise<JobOpening> {
    const response = await api.post(`${this.baseUrl}/job-openings`, jobOpening)
    return response.data
  }

  async getCandidates(params?: {
    job_opening_id?: string
    status?: string
  }): Promise<{ data: Candidate[] }> {
    const response = await api.get(`${this.baseUrl}/candidates`, { params })
    return response
  }

  async createCandidate(candidate: Omit<Candidate, 'id' | 'tenant_id' | 'created_at' | 'updated_at'>): Promise<Candidate> {
    const response = await api.post(`${this.baseUrl}/candidates`, candidate)
    return response.data
  }

  async getInterviews(params?: {
    candidate_id?: string
    interviewer_id?: string
    status?: string
  }): Promise<{ data: Interview[] }> {
    const response = await api.get(`${this.baseUrl}/interviews`, { params })
    return response
  }

  async createInterview(interview: Omit<Interview, 'id' | 'tenant_id' | 'created_at' | 'updated_at'>): Promise<Interview> {
    const response = await api.post(`${this.baseUrl}/interviews`, interview)
    return response.data
  }

  // Analytics
  async getHRAnalytics(): Promise<{ data: HRAnalytics }> {
    const response = await api.get(`${this.baseUrl}/analytics`)
    return response
  }

  // Employee Self-Service
  async getEmployeeDashboard(employeeId: string): Promise<{ data: EmployeeDashboard }> {
    const response = await api.get(`${this.baseUrl}/employees/${employeeId}/dashboard`)
    return response
  }

  // Utility Methods
  formatCurrency(amount: number): string {
    return new Intl.NumberFormat('en-US', {
      style: 'currency',
      currency: 'USD'
    }).format(amount)
  }

  formatDate(dateString: string): string {
    return new Date(dateString).toLocaleDateString()
  }

  getEmploymentTypeLabel(type: string): string {
    const labels: Record<string, string> = {
      FULL_TIME: 'Full Time',
      PART_TIME: 'Part Time',
      CONTRACT: 'Contract',
      TEMPORARY: 'Temporary',
      INTERN: 'Intern'
    }
    return labels[type] || type
  }

  getStatusColor(status: string): string {
    const colors: Record<string, string> = {
      ACTIVE: 'success',
      INACTIVE: 'secondary',
      PENDING: 'warning',
      APPROVED: 'success',
      REJECTED: 'danger',
      DRAFT: 'info',
      COMPLETED: 'success'
    }
    return colors[status] || 'secondary'
  }
}

export const hrmService = new HRMService()
export default hrmService